### DISCUSSION TITLE
"Error updating branch" is a WARN, but IMO should fail renovate execution

### DISCUSSION THREAD

**Discussion by wleese** [renovatebot/renovate#30197] [Request Help] [answered]
### What would you like help with?

I think I found a bug

### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us which platform (GitHub, GitLab, etc) and which version of Renovate.

GitLab

### Please tell us more about your question or problem

We recently set up restrictions on what branch names are permitted in GitLab.
This resulted in breaking our Renovate job, without us noticing.

The job outputs:

```
WARN: Error updating branch (repository=aaaa/aaaa-bb/aaaa, baseBranch=13.x, branch=renovate/13.x-cc-dd.version)
       "err": {
         "task": {
           "commands": [
             "push",
             "origin",
             "renovate/13.x-cc-dd.version:renovate/13.x-cc-dd.version",
             "--force-with-lease",
             "-u",
             "--no-verify",
             "--verbose",
             "--porcelain"
           ],
           "format": "utf-8",
           "parser": "[function]"
         },
         "message": "To https://gitlab.internal/aaaa/aaaa-bb/aaaa.git\n!\trefs/heads/renovate/13.x-cc-dd.version:refs/heads/renovate/13.x-cc-dd.version\t[remote rejected] (pre-receive hook declined)\nDone\nPushing to https://gitlab.internal/aaaa/aaaa-bb/aaaa.git\nPOST git-receive-pack (846 bytes)\nremote: GitLab: Branch name 'renovate/13.x-cc-dd.version' does not follow the pattern '^(\\d+\\.x|\\d+.\\d+.x|docs-latest-release-\\d+|release\\/.+|feature\\/.+)$'\nerror: failed to push some refs to 'https://gitlab.internal/aaaa/aaaa-bb/aaaa.git'\n",
         "stack": "Error: To https://**redacted**@3.25.0/node_modules/simple-git/src/lib/plugins/error-detection.plugin.ts:42:29)\n    at PluginStore.exec (/usr/local/renovate/node_modules/.pnpm/simple-git@3.25.0/node_modules/simple-git/src/lib/plugins/plugin-store.ts:54:29)\n    at /usr/local/renovate/node_modules/.pnpm/simple-git@3.25.0/node_modules/simple-git/src/lib/runners/git-executor-chain.ts:124:42\n    at new Promise (<anonymous>)\n    at GitExecutorChain.handleTaskData (/usr/local/renovate/node_modules/.pnpm/simple-git@3.25.0/node_modules/simple-git/src/lib/runners/git-executor-chain.ts:121:14)\n    at GitExecutorChain.<anonymous> (/usr/local/renovate/node_modules/.pnpm/simple-git@3.25.0/node_modules/simple-git/src/lib/runners/git-executor-chain.ts:97:40)\n    at Generator.next (<anonymous>)\n    at fulfilled (/usr/local/renovate/node_modules/.pnpm/simple-git@3.25.0/node_modules/simple-git/dist/cjs/index.js:52:24)\n    at processTicksAndRejections (node:internal/process/task_queues:95:5)"
       }
```

This effect of this warning is that for this dependency, no Merge Request will be created. The execution does not fail, so without checking the logs every day, it is impossible to know if there is a dependency update, or a problem with the GitLab configuration.

I was unable to find a configuration option that would allow me to ensure that these situations would have renovate exit with an error.

### Logs (if relevant)

<details><summary>Logs</summary>

```

Replace this text with your logs, between the starting and ending triple backticks

```

</details>

**Comment by RahulGautamSingh** [comment]
Have you tried https://docs.renovatebot.com/configuration-options/#loglevelremap?

**Comment by wleese** [comment]
I'll give it a try, but I do not consider this a valid solution at the moment.
If it works, Loglevel remapping will be brittle.
There's also the question of if Renovate cannot create/update a branch, if that simply shouldn't be considered a fatal error.
I would think so.

**Reply by RahulGautamSingh** [reply]
My thoughts: (you could wait for a maintainers comment if these do not asnwer your qn,)

1. If you have set up some rules for branch naming then there are config options(eg. `branchPrefix`) which allow you to modify the branch names created by the bot. Though it is limited you can use that.

2. We don't error and stop on one branch failing because we could be creating/updating multiple branches in that run and exiting cause one branch failed will result in all those PRs being halted.

> If it works, Loglevel remapping will be brittle.

How will it be brittle?

**Reply by wleese** [reply]
Log mapping would depend on the error message being the same. Future updates could change this and silently break this bit of configuration.

**Comment by rarkins** [comment]
Errors are meant for when Renovate is at fault, e.g. a bug. This is a case where the user is at fault, so WARN is correct. The logLevelRemap suggestion is the perfect way to handle when someone has a different opinion/need.

**Reply by wleese** [reply]
Is it Renovates stance that as long as it could run until completion, that any issue with e.g. a source management tool is not reason to exit with a failure?

In our use case, and I imagine other peoples, Renovate runs as a job. Its task is to keep dependencies up to date (or help us with that). The creation of Merge Requests (/ PRs) is behavior that I expect to be associated with Renovate successfully running. Likewise the non-creation of Merge Requests but the successful execution of Renovate, I associate with "there are no dependency updates".

When you say "the user is at fault", I would like to argue that regardless of how poorly the user configures the environment Renovate runs in, Renovate should still "care" about the user experience. Even if one would take the stance "the user is to blame, we're not going to fix it because they can", doesn't mean the user actually is at fault here. GitLab could be managed by an external party / department, or the public version might make a breaking change.

Replying to @RahulGautamSingh's "We don't error and stop on one branch failing": this is definitely not what I'm asking. I understand Renovate should run until completion, so that other work can be completed.

**Comment by wleese** [comment]
See my last comment (https://github.com/renovatebot/renovate/discussions/30197#discussioncomment-10061591), is there no interest in improving the situation here?

**Reply by rarkins** [reply, marked as answer]
No. We have decided it's a warn and given ability for you to remap it.

**Comment by breun** [comment]
The current behavior has caused a lot of failures to create merge requests (for instance due to branch protection rule changes) to go unnoticed for quite a while for my team, because we expected that a successful run of Renovate would mean that it created merge requests for all identified updates. It surprised us to learn that this is not necessarily the case and that a successful Renovate execution can mean that merge request creation failed for some or all of the identified updates.
